### PR TITLE
PP-12758 Pipeline to run vulnerability scan

### PR DESCRIPTION
**Pull request by kbottla**
# WHAT
- pipeline to run internal vulnerability scan

Concourse job https://pay-cd.deploy.payments.service.gov.uk/teams/pay-deploy/pipelines/internal-vulnerability-scan/jobs/run-vulnerability-scan/builds/1

### CODE REVIEW

**Comment by alphagov-pay-ci-concourse** [comment]
**No YAML differences detected between PR and master pkl files in Concourse Team pay-dev**

**Comment by alphagov-pay-ci-concourse** [comment]
# Changes for pay-deploy/internal-vulnerability-scan.pkl

<details><summary>Diff of YAML generated from pay-deploy/internal-vulnerability-scan.pkl</summary>

```diff
--- pay-ci/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
+++ pkl-pipeline-pr/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
@@ -0,0 +1,104 @@
+jobs:
+- name: update-pipeline
+  plan:
+  - get: pipeline-source
+    trigger: true
+  - task: generate-yaml-from-pkl
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source:
+          repository: governmentdigitalservice/pay-concourse-runner
+          tag: latest
+      inputs:
+      - name: pipeline-source
+      outputs:
+      - name: pipeline-source
+      run:
+        path: ash
+        args:
+        - -c
+        - mkdir -p 'pay-deploy' && pkl eval 'pay-deploy/internal-vulnerability-scan.pkl' > 'pay-deploy/internal-vulnerability-scan.yaml'
+        dir: pipeline-source/ci/pkl-pipelines/
+  - set_pipeline: internal-vulnerability-scan
+    file: pipeline-source/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yaml
+- name: run-vulnerability-scan
+  plan:
+  - get: pay-ci
+  - task: assume-read-from-staging-ecr-role
+    file: pay-ci/ci/tasks/assume-role.yml
+    params:
+      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
+      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
+    output_mapping:
+      assume-role: assume-read-from-staging-ecr-role
+  - load_var: assume-read-from-staging-ecr-role
+    file: assume-read-from-staging-ecr-role/assume-role.json
+    format: json
+  - task: run-vulnerability-scan
+    file: pay-ci/ci/tasks/run-vulnerability-scan.yml
+    privileged: true
+    params:
+      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
+      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
+      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
+      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
+      AWS_DEFAULT_REGION: eu-west-1
+      DOCKERHUB_USERNAME: ((docker-username))
+      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
+    input_mapping:
+      pay-ci: pay-ci
+  on_success:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+  on_failure:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+resources:
+- name: pipeline-source
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+    paths:
+    - ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.pkl
+    - ci/pkl-pipelines/common/**
+  icon: github
+- name: pay-ci
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+  icon: github
+- name: slack-notification
+  type: slack-notification
+  source:
+    url: https://hooks.slack.com/services/((slack-notification-secret))
+resource_types:
+- name: pull-request
+  type: registry-image
+  source:
+    repository: teliaoss/github-pr-resource
+    tag: v0.23.0
+- name: slack-notification
+  type: docker-image
+  source:
+    repository: cfcommunity/slack-notification-resource
+    tag: latest
```
</details>


<details><summary>Concourse set-pipeline dry-run for `pay-deploy/internal-vulnerability-scan.pkl`</summary>

```diff
resources:
  resource pay-ci has changed:
  icon: github
  name: pay-ci
  source:
-   branch: pp_12758_pipeline_for_vul_scan
+   branch: master
    password: ((github-access-token))
    uri: https://github.com/alphagov/pay-ci
    username: alphagov-pay-ci-concourse
  type: git
  
pipeline name: internal-vulnerability-scan

Dry-run mode was set, exiting.
```

**Comment by alphagov-pay-ci-concourse** [comment]
The following files had no diff in the YAML between the main branch and this PR:

```
pay-deploy/concourse-runner.yml
pay-deploy/deploy-to-production.yml
pay-deploy/deploy-to-staging.yml
pay-deploy/detect-secrets.yml
pay-deploy/infra-drift-detector.yml
pay-deploy/init-lock-pools.yml
pay-deploy/pact-broker.yml
pay-deploy/pay-js.yml
pay-deploy/pkl-pipeline-changes.yml
pay-deploy/pool-resource.yml
pay-deploy/prometheus-pushgateway.yml
```

File: ci/scripts/run-vulnerability-scan/scan.sh [hunk 1]
```diff
@@ -0,0 +1,63 @@
+#!/bin/bash
```
[the remark this answers is not in the document]
Comment (kbottla) [reply ci/scripts/run-vulnerability-scan/scan.sh]: 👍🏼 updated

**Comment by alphagov-pay-ci-concourse** [comment]
# Changes for pay-deploy/internal-vulnerability-scan.pkl

<details><summary>Diff of YAML generated from pay-deploy/internal-vulnerability-scan.pkl</summary>

```diff
--- pay-ci/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
+++ pkl-pipeline-pr/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
@@ -0,0 +1,104 @@
+jobs:
+- name: update-pipeline
+  plan:
+  - get: pipeline-source
+    trigger: true
+  - task: generate-yaml-from-pkl
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source:
+          repository: governmentdigitalservice/pay-concourse-runner
+          tag: latest
+      inputs:
+      - name: pipeline-source
+      outputs:
+      - name: pipeline-source
+      run:
+        path: ash
+        args:
+        - -c
+        - mkdir -p 'pay-deploy' && pkl eval 'pay-deploy/internal-vulnerability-scan.pkl' > 'pay-deploy/internal-vulnerability-scan.yaml'
+        dir: pipeline-source/ci/pkl-pipelines/
+  - set_pipeline: internal-vulnerability-scan
+    file: pipeline-source/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yaml
+- name: run-vulnerability-scan
+  plan:
+  - get: pay-ci
+  - task: assume-read-from-staging-ecr-role
+    file: pay-ci/ci/tasks/assume-role.yml
+    params:
+      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
+      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
+    output_mapping:
+      assume-role: assume-read-from-staging-ecr-role
+  - load_var: assume-read-from-staging-ecr-role
+    file: assume-read-from-staging-ecr-role/assume-role.json
+    format: json
+  - task: run-vulnerability-scan
+    file: pay-ci/ci/tasks/run-vulnerability-scan.yml
+    privileged: true
+    params:
+      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
+      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
+      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
+      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
+      AWS_DEFAULT_REGION: eu-west-1
+      DOCKERHUB_USERNAME: ((docker-username))
+      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
+    input_mapping:
+      pay-ci: pay-ci
+  on_success:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+  on_failure:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+resources:
+- name: pipeline-source
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+    paths:
+    - ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.pkl
+    - ci/pkl-pipelines/common/**
+  icon: github
+- name: pay-ci
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+  icon: github
+- name: slack-notification
+  type: slack-notification
+  source:
+    url: https://hooks.slack.com/services/((slack-notification-secret))
+resource_types:
+- name: pull-request
+  type: registry-image
+  source:
+    repository: teliaoss/github-pr-resource
+    tag: v0.23.0
+- name: slack-notification
+  type: docker-image
+  source:
+    repository: cfcommunity/slack-notification-resource
+    tag: latest
```
</details>


<details><summary>Concourse set-pipeline dry-run for `pay-deploy/internal-vulnerability-scan.pkl`</summary>

```diff
resources:
  resource pay-ci has changed:
  icon: github
  name: pay-ci
  source:
-   branch: pp_12758_pipeline_for_vul_scan
+   branch: master
    password: ((github-access-token))
    uri: https://github.com/alphagov/pay-ci
    username: alphagov-pay-ci-concourse
  type: git
  
pipeline name: internal-vulnerability-scan

Dry-run mode was set, exiting.
```

**Comment by alphagov-pay-ci-concourse** [comment]
The following files had no diff in the YAML between the main branch and this PR:

```
pay-deploy/concourse-runner.yml
pay-deploy/deploy-to-production.yml
pay-deploy/deploy-to-staging.yml
pay-deploy/detect-secrets.yml
pay-deploy/infra-drift-detector.yml
pay-deploy/init-lock-pools.yml
pay-deploy/pact-broker.yml
pay-deploy/pay-js.yml
pay-deploy/pkl-pipeline-changes.yml
pay-deploy/pool-resource.yml
pay-deploy/prometheus-pushgateway.yml
```

**Comment by alphagov-pay-ci-concourse** [comment]
**No YAML differences detected between PR and master pkl files in Concourse Team pay-dev**

**Comment by jfharden** [comment]
@kbottla The job didn't actually work, if you check the output `Scanning image 888564216586.dkr.ecr.eu-west-1.amazonaws.com/govukpay/adminusers:1831-release
    i Failed to connect to Docker Engine. Please make sure Docker Engine is running to access local images.` it's trying to talk to the docker daemon which isn't running. It should just hard fail! But it carries on like everything is ok! (wild behaviour)
    
 We have a helper script in concourse runner that will start it up, the other scripts we use it in have this snippet in
  
 ```
    source /docker-helpers.sh
    start_docker

    function cleanup {
      echo "CLEANUP TRIGGERED"
      clean_docker
      stop_docker
      echo "CLEANUP COMPLETE"
    }

    trap cleanup EXIT
```

**Comment by jfharden** [comment]
@kbottla This is really weird, if I run docker scout cves locally without docker running it exits with a fail 🤔 

```
15:36 : GDS10799 : no_aws : (master) : [~/gitdevel/github.com/alphagov/pay-ci/ci]
♥ $ docker scout cves 223851549868.dkr.ecr.eu-west-1.amazonaws.com/govukpay/docker-nginx-proxy:40-release
    i New version 1.10.0 available (installed version is 1.9.3) at https://github.com/docker/scout-cli
          i Failed to connect to Docker Engine. Please make sure Docker Engine is running to access local images.
    ✗ Pull failed
ERROR   Status: failed to get image 223851549868.dkr.ecr.eu-west-1.amazonaws.com/govukpay/docker-nginx-proxy:40-release: failed to pull image 223851549868.dkr.ecr.eu-west-1.amazonaws.com/govukpay/docker-nginx-proxy:40-release: GET https://223851549868.dkr.ecr.eu-west-1.amazonaws.com/v2/govukpay/docker-nginx-proxy/manifests/40-release: DENIED: Your authorization token has expired. Reauthenticate and try again., Code: 1
15:39 : GDS10799 : no_aws : (master) : [~/gitdevel/github.com/alphagov/pay-ci/ci]
♥ $ echo $?
1
```

I wonder why it doesn't fail like that in concourse 🤔

**Comment by kbottla** [comment]
@jfharden yeah, it is weird. Although there are errors on concourse, the report is being generated correctly. I will look into adding the docker start-up script.

**Comment by alphagov-pay-ci-concourse** [comment]
# Changes for pay-deploy/internal-vulnerability-scan.pkl

<details><summary>Diff of YAML generated from pay-deploy/internal-vulnerability-scan.pkl</summary>

```diff
--- pay-ci/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
+++ pkl-pipeline-pr/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
@@ -0,0 +1,104 @@
+jobs:
+- name: update-pipeline
+  plan:
+  - get: pipeline-source
+    trigger: true
+  - task: generate-yaml-from-pkl
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source:
+          repository: governmentdigitalservice/pay-concourse-runner
+          tag: latest
+      inputs:
+      - name: pipeline-source
+      outputs:
+      - name: pipeline-source
+      run:
+        path: ash
+        args:
+        - -c
+        - mkdir -p 'pay-deploy' && pkl eval 'pay-deploy/internal-vulnerability-scan.pkl' > 'pay-deploy/internal-vulnerability-scan.yaml'
+        dir: pipeline-source/ci/pkl-pipelines/
+  - set_pipeline: internal-vulnerability-scan
+    file: pipeline-source/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yaml
+- name: run-vulnerability-scan
+  plan:
+  - get: pay-ci
+  - task: assume-read-from-staging-ecr-role
+    file: pay-ci/ci/tasks/assume-role.yml
+    params:
+      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
+      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
+    output_mapping:
+      assume-role: assume-read-from-staging-ecr-role
+  - load_var: assume-read-from-staging-ecr-role
+    file: assume-read-from-staging-ecr-role/assume-role.json
+    format: json
+  - task: run-vulnerability-scan
+    file: pay-ci/ci/tasks/run-vulnerability-scan.yml
+    privileged: true
+    params:
+      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
+      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
+      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
+      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
+      AWS_DEFAULT_REGION: eu-west-1
+      DOCKERHUB_USERNAME: ((docker-username))
+      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
+    input_mapping:
+      pay-ci: pay-ci
+  on_success:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+  on_failure:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+resources:
+- name: pipeline-source
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+    paths:
+    - ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.pkl
+    - ci/pkl-pipelines/common/**
+  icon: github
+- name: pay-ci
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+  icon: github
+- name: slack-notification
+  type: slack-notification
+  source:
+    url: https://hooks.slack.com/services/((slack-notification-secret))
+resource_types:
+- name: pull-request
+  type: registry-image
+  source:
+    repository: teliaoss/github-pr-resource
+    tag: v0.23.0
+- name: slack-notification
+  type: docker-image
+  source:
+    repository: cfcommunity/slack-notification-resource
+    tag: latest
```
</details>


<details><summary>Concourse set-pipeline dry-run for `pay-deploy/internal-vulnerability-scan.pkl`</summary>

```diff
resources:
  resource pay-ci has changed:
  icon: github
  name: pay-ci
  source:
-   branch: pp_12758_pipeline_for_vul_scan
+   branch: master
    password: ((github-access-token))
    uri: https://github.com/alphagov/pay-ci
    username: alphagov-pay-ci-concourse
  type: git
  
jobs:
  job run-vulnerability-scan has changed:
  name: run-vulnerability-scan
  on_failure:
    attempts: 10
    params:
-     channel: '#kb-test-support-notifications'
+     channel: '#govuk-pay-starling'
      icon_emoji: ':concourse:'
      silent: true
      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse
        build #$BUILD_NAME>'
      username: pay-concourse
    put: slack-notification
  on_success:
    attempts: 10
    params:
-     channel: '#kb-test-support-notifications'
+     channel: '#govuk-pay-starling'
      icon_emoji: ':concourse:'
      silent: true
      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse
        build #$BUILD_NAME>'
      username: pay-concourse
    put: slack-notification
  plan:
  - get: pay-ci
  - file: pay-ci/ci/tasks/assume-role.yml
    output_mapping:
      assume-role: assume-read-from-staging-ecr-role
    params:
      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
    task: assume-read-from-staging-ecr-role
  - file: assume-read-from-staging-ecr-role/assume-role.json
    format: json
    load_var: assume-read-from-staging-ecr-role
  - file: pay-ci/ci/tasks/run-vulnerability-scan.yml
    input_mapping:
      pay-ci: pay-ci
    params:
      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
      AWS_DEFAULT_REGION: eu-west-1
      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
      DOCKERHUB_USERNAME: ((docker-username))
      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
    privileged: true
    task: run-vulnerability-scan
  
pipeline name: internal-vulnerability-scan

Dry-run mode was set, exiting.
```

**Comment by alphagov-pay-ci-concourse** [comment]
The following files had no diff in the YAML between the main branch and this PR:

```
pay-deploy/concourse-runner.yml
pay-deploy/deploy-to-production.yml
pay-deploy/deploy-to-staging.yml
pay-deploy/detect-secrets.yml
pay-deploy/infra-drift-detector.yml
pay-deploy/init-lock-pools.yml
pay-deploy/pact-broker.yml
pay-deploy/pay-js.yml
pay-deploy/pkl-pipeline-changes.yml
pay-deploy/pool-resource.yml
pay-deploy/prometheus-pushgateway.yml
```

**Comment by alphagov-pay-ci-concourse** [comment]
# Changes for pay-deploy/internal-vulnerability-scan.pkl

<details><summary>Diff of YAML generated from pay-deploy/internal-vulnerability-scan.pkl</summary>

```diff
--- pay-ci/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
+++ pkl-pipeline-pr/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yml
@@ -0,0 +1,104 @@
+jobs:
+- name: update-pipeline
+  plan:
+  - get: pipeline-source
+    trigger: true
+  - task: generate-yaml-from-pkl
+    config:
+      platform: linux
+      image_resource:
+        type: registry-image
+        source:
+          repository: governmentdigitalservice/pay-concourse-runner
+          tag: latest
+      inputs:
+      - name: pipeline-source
+      outputs:
+      - name: pipeline-source
+      run:
+        path: ash
+        args:
+        - -c
+        - mkdir -p 'pay-deploy' && pkl eval 'pay-deploy/internal-vulnerability-scan.pkl' > 'pay-deploy/internal-vulnerability-scan.yaml'
+        dir: pipeline-source/ci/pkl-pipelines/
+  - set_pipeline: internal-vulnerability-scan
+    file: pipeline-source/ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.yaml
+- name: run-vulnerability-scan
+  plan:
+  - get: pay-ci
+  - task: assume-read-from-staging-ecr-role
+    file: pay-ci/ci/tasks/assume-role.yml
+    params:
+      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
+      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
+    output_mapping:
+      assume-role: assume-read-from-staging-ecr-role
+  - load_var: assume-read-from-staging-ecr-role
+    file: assume-read-from-staging-ecr-role/assume-role.json
+    format: json
+  - task: run-vulnerability-scan
+    file: pay-ci/ci/tasks/run-vulnerability-scan.yml
+    privileged: true
+    params:
+      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
+      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
+      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
+      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
+      AWS_DEFAULT_REGION: eu-west-1
+      DOCKERHUB_USERNAME: ((docker-username))
+      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
+    input_mapping:
+      pay-ci: pay-ci
+  on_success:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+  on_failure:
+    attempts: 10
+    put: slack-notification
+    params:
+      channel: '#govuk-pay-starling'
+      silent: true
+      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse build #$BUILD_NAME>'
+      icon_emoji: ':concourse:'
+      username: pay-concourse
+resources:
+- name: pipeline-source
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+    paths:
+    - ci/pkl-pipelines/pay-deploy/internal-vulnerability-scan.pkl
+    - ci/pkl-pipelines/common/**
+  icon: github
+- name: pay-ci
+  type: git
+  source:
+    uri: https://github.com/alphagov/pay-ci
+    branch: master
+    username: alphagov-pay-ci-concourse
+    password: ((github-access-token))
+  icon: github
+- name: slack-notification
+  type: slack-notification
+  source:
+    url: https://hooks.slack.com/services/((slack-notification-secret))
+resource_types:
+- name: pull-request
+  type: registry-image
+  source:
+    repository: teliaoss/github-pr-resource
+    tag: v0.23.0
+- name: slack-notification
+  type: docker-image
+  source:
+    repository: cfcommunity/slack-notification-resource
+    tag: latest
```
</details>


<details><summary>Concourse set-pipeline dry-run for `pay-deploy/internal-vulnerability-scan.pkl`</summary>

```diff
resources:
  resource pay-ci has changed:
  icon: github
  name: pay-ci
  source:
-   branch: pp_12758_pipeline_for_vul_scan
+   branch: master
    password: ((github-access-token))
    uri: https://github.com/alphagov/pay-ci
    username: alphagov-pay-ci-concourse
  type: git
  
jobs:
  job run-vulnerability-scan has changed:
  name: run-vulnerability-scan
  on_failure:
    attempts: 10
    params:
-     channel: '#kb-test-support-notifications'
+     channel: '#govuk-pay-starling'
      icon_emoji: ':concourse:'
      silent: true
      text: ':red-circle: Failed to run vulnerability scan - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse
        build #$BUILD_NAME>'
      username: pay-concourse
    put: slack-notification
  on_success:
    attempts: 10
    params:
-     channel: '#kb-test-support-notifications'
+     channel: '#govuk-pay-starling'
      icon_emoji: ':concourse:'
      silent: true
      text: ':green-circle: Internal vulnerability scan has been completed - <https://pay-cd.deploy.payments.service.gov.uk/teams/$BUILD_TEAM_NAME/pipelines/$BUILD_PIPELINE_NAME/jobs/$BUILD_JOB_NAME/builds/$BUILD_NAME|Concourse
        build #$BUILD_NAME>'
      username: pay-concourse
    put: slack-notification
  plan:
  - get: pay-ci
  - file: pay-ci/ci/tasks/assume-role.yml
    output_mapping:
      assume-role: assume-read-from-staging-ecr-role
    params:
      AWS_ROLE_ARN: arn:aws:iam::((pay_aws_staging_account_id)):role/concourse
      AWS_ROLE_SESSION_NAME: read-from-ecr-in-staging
    task: assume-read-from-staging-ecr-role
  - file: assume-read-from-staging-ecr-role/assume-role.json
    format: json
    load_var: assume-read-from-staging-ecr-role
  - file: pay-ci/ci/tasks/run-vulnerability-scan.yml
    input_mapping:
      pay-ci: pay-ci
    params:
      AWS_ACCESS_KEY_ID: ((.:assume-read-from-staging-ecr-role.AWS_ACCESS_KEY_ID))
      AWS_DEFAULT_REGION: eu-west-1
      AWS_SECRET_ACCESS_KEY: ((.:assume-read-from-staging-ecr-role.AWS_SECRET_ACCESS_KEY))
      AWS_SESSION_TOKEN: ((.:assume-read-from-staging-ecr-role.AWS_SESSION_TOKEN))
      DOCKERHUB_ACCESS_TOKEN: ((docker-access-token))
      DOCKERHUB_USERNAME: ((docker-username))
      ECR_REGISTRY: ((pay_aws_staging_account_id)).dkr.ecr.eu-west-1.amazonaws.com
    privileged: true
    task: run-vulnerability-scan
  
pipeline name: internal-vulnerability-scan

Dry-run mode was set, exiting.
```

**Comment by alphagov-pay-ci-concourse** [comment]
The following files had no diff in the YAML between the main branch and this PR:

```
pay-deploy/concourse-runner.yml
pay-deploy/deploy-to-production.yml
pay-deploy/deploy-to-staging.yml
pay-deploy/detect-secrets.yml
pay-deploy/infra-drift-detector.yml
pay-deploy/init-lock-pools.yml
pay-deploy/pact-broker.yml
pay-deploy/pay-js.yml
pay-deploy/pkl-pipeline-changes.yml
pay-deploy/pool-resource.yml
pay-deploy/prometheus-pushgateway.yml
```

**Comment by kbottla** [comment]
@jfharden Added docker helper script and there are no more errors. Latest build https://pay-cd.deploy.payments.service.gov.uk/teams/pay-deploy/pipelines/internal-vulnerability-scan/jobs/run-vulnerability-scan/builds/6

**Review by jfharden — APPROVED** [review]

**Pull request merged**